core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java

```diff
@@ -0,0 +1,227 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.retry;
+
+import static com.linecorp.armeria.common.util.Functions.voidFunction;
+
+import java.time.Duration;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.client.Client;
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.common.Flags;
+import com.linecorp.armeria.common.http.DefaultHttpRequest;
+import com.linecorp.armeria.common.http.DefaultHttpResponse;
+import com.linecorp.armeria.common.http.HttpHeaderNames;
+import com.linecorp.armeria.common.http.HttpHeaders;
+import com.linecorp.armeria.common.http.HttpObject;
+import com.linecorp.armeria.common.http.HttpRequest;
+import com.linecorp.armeria.common.http.HttpResponse;
+import com.linecorp.armeria.common.http.HttpResponseDuplicator;
+import com.linecorp.armeria.common.http.HttpStatus;
+import com.linecorp.armeria.common.http.HttpStatusClass;
+
+import io.netty.channel.EventLoop;
+
+/**
+ * A {@link Client} decorator that handles failures of an invocation and retries HTTP requests.
+ */
+public class RetryingHttpClient extends RetryingClient<HttpRequest, HttpResponse> {
+
+    private static final Logger logger = LoggerFactory.getLogger(RetryingHttpClient.class);
+
+    private final boolean useRetryAfter;
+
+    /**
+     * Creates a new {@link Client} decorator that handles failures of an invocation and retries HTTP requests.
+     */
+    public static Function<Client<HttpRequest, HttpResponse>, RetryingHttpClient>
+    newDecorator(RetryRequestStrategy<HttpRequest, HttpResponse> retryRequestStrategy,
+                 Supplier<? extends Backoff> backoffSupplier) {
+        return delegate ->
+                new RetryingHttpClient(delegate, retryRequestStrategy, backoffSupplier,
+                                       Flags.defaultBackoffMaxAttempts(), false);
+    }
+
+    /**
+     * Creates a new instance that decorates the specified {@link Client}.
+     */
+    public RetryingHttpClient(Client<HttpRequest, HttpResponse> delegate,
```

Comment:
I wonder if this could be package-private now that we have a builder. (and this class could be final then.)